compiler/rustc_hir_analysis/src/check/intrinsic.rs

```diff
@@ -222,6 +222,18 @@ pub fn check_intrinsic_type(tcx: TyCtxt<'_>, it: &hir::ForeignItem<'_>) {
                 ],
                 tcx.mk_ptr(ty::TypeAndMut { ty: param(0), mutbl: hir::Mutability::Not }),
             ),
+            sym::option_payload_ptr => {
+                let Some(option_def_id) = tcx.get_diagnostic_item(sym::Option) else { bug!() };
+                let p0 = param(0);
+                (
+                    1,
+                    vec![tcx.mk_ptr(ty::TypeAndMut {
+                        ty: tcx.type_of(option_def_id).subst(tcx, &[p0.into()]),
```

Comment:
Maybe something like `tcx.mk_adt(tcx.adt_def(option_def_id), tcx.check_and_mk_substs(option_def_id, [p0]))`

Reply:
Again, `check_and_mk_substs` is private to `ty`, but I can just create the substs from iter.